src/main/scala/com/snowflake/snowpark/internal/ServerConnection.scala

```diff
@@ -857,20 +871,24 @@ private[snowpark] class ServerConnection(
 
         logDebug(s"""execute plan in async mode:
                    |----------SNOW-----------
+                   |
                    |$plan
                    |-------------------------
                    |""".stripMargin)
 
         // use try finally to ensure postActions is always run
-        val statement = connection.createStatement()
+        val queries = plan.queries.map(_.sql)
+        val multipleStatements = queries.mkString("; ")
+        val statement = connection.prepareStatement(multipleStatements)
         try {
-          val queries = plan.queries.map(_.sql)
-          val multipleStatements = queries.mkString("; ")
+          // Note binding parameters only supported for single query
+          val bindingParameters =
+            if (plan.queries.length == 1) plan.queries.last.params else Seq()
```

Comment:
If it doesn't support multiple statements, let's throw a user exception here.
you can find example from there https://github.com/snowflakedb/snowpark-java-scala/blob/main/src/main/scala/com/snowflake/snowpark/internal/ErrorMessage.scala

I checked the code, the only one usage is caching dataframe. So user can't cache dataframe if invoke binding.
I think we can improve it later. Now we cache dataframe by `Create Table` + `Insert Into table`, these two queries can be combined into one. Anyway, let's don't do it now, just throw one exception.

Reply:
Note I had to consoldiate some of the Query.apply methods as scala couldn't resolve them otherwise, seems like an limitation with scala.